llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp

```diff
@@ -0,0 +1,125 @@
+//===-- LowerWGLocalMemory.cpp - SYCL kernel local memory allocation pass -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass replaces calls to __sycl_allocateLocalMemory(Size, Alignment)
+// function with allocation of memory in local address space at the kernel
+// scope.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/LowerWGLocalMemory.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "LowerWGLocalMemory"
+
+static constexpr char SYCL_ALLOCLOCALMEM_CALL[] = "__sycl_allocateLocalMemory";
+static constexpr char LOCALMEMORY_GV_PREF[] = "WGLocalMem";
+
+namespace {
+class SYCLLowerWGLocalMemoryLegacy : public ModulePass {
```

Comment:
Not clear why this is a `ModulePass` rather than `FunctionPass` - the pass does not seem to perform any cross-function actions. If none - suggest to make it FunctionPass

Reply:
I was guided by FunctionPass restrictions from LLVM documentation (https://llvm.org/docs/WritingAnLLVMPass.html#the-functionpass-class):
> 3. Add or remove global variables from the current Module.